aws/adapter.go

```diff
@@ -568,14 +593,26 @@ func (a *Adapter) UpdateTargetGroupsAndAutoScalingGroups(stacks []*Stack, proble
 		return
 	}
 
+	// split the full list into relevant TG types
+	targetTypesARNs, err := categorizeTargetTypeInstance(a.elbv2, targetGroupARNs)
```

Comment:
`categorizeTargetTypeInstance` fetches all target groups which we do once again inside of `updateTargetGroupsForAutoScalingGroup` which fells unaligned. 

The `targetTypesARNs` contains only exiting target groups out of `targetGroupARNs` which is not correct - the `updateTargetGroupsForAutoScalingGroup` also handles the case when TG from `targetGroupARNs` does not exist.

The proposed approach with CNI updates is different from the existing design: controller runs a loop which calls `UpdateTargetGroupsAndAutoScalingGroups` on every iteration.

I'm wondering if we could keep it simple like it is, e.g. controller calls `UpdateTargetGroupsAndAutoScalingGroups`, it fetches pods and instances and updates target groups accordingly, all synchronous just like everything else.

Reply:
`updateTargetGroupsForAutoScalingGroup` is built with the assumption there is only one type of TG namely `instance`. However, as we got two types including `ip` we need to filter / categorize the set down respectively, done via `categorizeTargetTypeInstance`.
From that point I don't grasp how this is wrong @AlexanderYastrebov - it should leave the default flow untouched. I also hesitate to rewrite the existing functions, which might be an idea for the future though.

Reply:
I mean that there were several improvements like #127 or #436 over the time that made controller more resilient to the external, possibly accidental, changes. These improvements were not invented out of nothing but rather from real problems and incidents. Pre-filtering stack target groups using existing target groups wipes those improvements. 

We may argue that controller resources should not be changed externally; if we could guarantee this then codebase would be even simpler. But for now I am not convinced we should change things by accidentally filtering via `categorizeTargetTypeInstance` (i.e. we do not leave default flow untouched).

Reply:
Excuse my ignorance and thanks for the additional context. It seems to me that the action of `categorizeTargetTypeInstance' is misunderstood here. There is no filtering happening in this function, just categorizing of the TG by types (i.e. `instance`, `ip`, `alb` etc.)
Therefor, the sum of `targetGroupARNs` equals to the sum of `targetTypesARNs[n]`.

Those mentioned improvements will still apply, because it is impossible that TG of any other type than `instance` are linked to an ASG group!
Hence it would be an error to send to `UpdateTargetGroupsAndAutoScalingGroups` a set of TG where any of them is not of type `instance`, it would simply fail. That's why we need to categorize the types here, and since this controller was ignorant of other TG types than `instance` before, this flow is not touched.

Reply:
@AlexanderYastrebov I guess you should add a proposal here, such that we can make the change happen.

Reply:
> There is no filtering happening in this function, just categorizing of the TG by types

The code adds to the resulting `targetTypes` only TGs returned by the `DescribeTargetGroupsPagesWithContext` and thus does not add non-existent/externally deleted target groups which still could be referenced by stacks.
 
I am wondering if we can use stack's `targetAccessModeCNI` to split their target groups by mode and get away without API call at all.

Reply:
> I am wondering if we can use stack's targetAccessModeCNI to split their target groups by mode and get away without API call at all.

I guess we could add mode as a CF template parameter and use it to set `aws.Stack` property in `mapToManagedStack`

Reply:
> please come up with an assisting code solution as it seems we are not progressing anywhere

We are not progressing on this point because the proposal changes existing flow:
1. create a stack
2. manually remove its TG
3. controler would reconcile and de-regester removed TG

With this proposal `targetTypesARNs[elbv2.TargetTypeEnumInstance]` passed to `updateTargetGroupsForAutoScalingGroup` and others will only contain existing TGs and thus removed TG would not be de-regestered from ASG.

Reply:
Let me try to picture a solution: `updateTargetGroupsForAutoScalingGroup` receives the original set of `targetGroupARNs` just minus those `targetTypesARNs[elbv2.TargetTypeEnumIp]`. That should include all your dirty TGs as well like before but filtered the type IP out.

Would you be happy with that @AlexanderYastrebov ?

Reply:
That sounds better and easier than tracking target mode as CF stack parameter.
```go
func (a *Adapter) UpdateTargetGroupsAndAutoScalingGroups(stacks []*Stack, problems *problem.List) {
	instanceTargetGroupARNs, ipTargetGroupARNs, err := getTargetGroups(stacks)
	if err != nil {
		problems.Add("failed to categorize Target Type Instance: %w", err)
		return
	}

	if a.TargetCNI.Enabled {
		a.TargetCNI.TargetGroupCh <- ipTargetGroupARNs
	}

	// don't do anything if there are no target groups
	if len(instanceTargetGroupARNs) == 0 {
		return
	}
	...
}
```

> Would you be happy with that

I would be happy to accept changes that will not create incidents/work and I try to foresee as much of the problems as possible during review.

Reply:
implemented